[dynamo] Support cuda stream and stream context reconstruction

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115410
* __->__ #117008
* #117006
* #116723

In dynamo, today, stream context reconstruction is using the generic reconstruction code, which was not correctly reconstruction the stream context with the user defined stream input. This PR:

1) Fixes the stream context to correctly reconstruct (as tested by test_cuda_stream_context_manager_graph_break_named) - we can now pass a stream context across a graph break and correctly enter it.

2) Add reconstruction support for streams by utilizing existing proxy-able object support (as tested by test_cuda_stream_context_manager_graph_break) - we can now pass a stream across graph breaks and use it, in the case of our test, we make a context with it and enter it. Note: This means that streams can be graph inputs and outputs now. 


cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

voznesenskym committed:
[dynamo] support cuda stream reconstruction, graph breaks and residuals

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/ctx_manager.py:
```diff
@@ -586,6 +589,12 @@ def call_method(
+        codegen.load_import_from("torch", "cuda.Stream")
+        codegen.extend_output([codegen.create_load_const(str(self.device))])
+        codegen.extend_output(create_call_function(1, False))
+        return []
```
Is it ever possible for the cuda stream here to be a preexisting cuda stream?

voznesenskym replied:
Sorry forgot to mark as draft - this PR aint done, no desc, etc

voznesenskym replied:
Yes - and this will re-create it. I spent a few hours testing it and it seems to functionally be the same thing. Would you prefer if we did not do this, and had reconstruct provide back the existing stream?

voznesenskym replied:
Actually, I'm wrong here.

voznesenskym committed:
Update on "[dynamo] support cuda stream reconstruction, graph breaks and residuals"




cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support cuda stream reconstruction, graph breaks and residuals"




cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] support cuda stream reconstruction, graph breaks and residuals"




cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -815,7 +815,6 @@ def popn(self, n: int) -> List[VariableTracker]:
 
     def LOAD_FAST(self, inst):
         name = inst.argval
-
```
self nit - undo spurious removal

voznesenskym committed:
Update on "[dynamo] Support cuda stream and stream context reconstruction"


In dynamo, today, stream context reconstruction is using the generic reconstruction code, which was not correctly reconstruction the stream context with the user defined stream input. This PR:

1) Fixes the stream context to correctly reconstruct (as tested by test_cuda_stream_context_manager_graph_break_named) - we can now pass a stream context across a graph break and correctly enter it.

2) Add reconstruction support for streams (as tested by test_cuda_stream_context_manager_graph_break) - we can now pass a stream across graph breaks and use it, in the case of our test, we make a context with it and enter it.


cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym committed:
Update on "[dynamo] Support cuda stream and stream context reconstruction"


In dynamo, today, stream context reconstruction is using the generic reconstruction code, which was not correctly reconstruction the stream context with the user defined stream input. This PR:

1) Fixes the stream context to correctly reconstruct (as tested by test_cuda_stream_context_manager_graph_break_named) - we can now pass a stream context across a graph break and correctly enter it.

2) Add reconstruction support for streams by utilizing existing proxy-able object support (as tested by test_cuda_stream_context_manager_graph_break) - we can now pass a stream across graph breaks and use it, in the case of our test, we make a context with it and enter it. Note: This means that streams can be graph inputs and outputs now. 


cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -570,14 +571,20 @@ def build_key_value(k, v):
+                value.device,
                 source=self.source,
             )
+            return self.tx.output.side_effects.track_object_existing(value, stream_var)
```
Maybe don't need this yet.

ezyang replied:
Do we support mutations on streams? I thought they're pretty immutable.

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -570,14 +571,20 @@ def build_key_value(k, v):
+            self.install_guards(GuardBuilder.ID_MATCH)
+            stream_source = AttrSource(self.source, "stream")
+            stream_var = VariableBuilder(self.tx, stream_source)(value.stream)
+            return StreamContextVariable.create_from_stream(self.tx, stream_var)
```
This is apparently preexisting, so no action needed, but it's weird to me that we need a dedicated StreamContextVariable, it seems like a pretty standard context manager around stream setting apis, it seems to me that we should just inline into it.

ezyang commented on torch/_dynamo/variables/ctx_manager.py:
```diff
@@ -521,24 +521,28 @@ def exit(self, tx, *args):
+        codegen.load_import_from("torch", "cuda.stream")
+        assert len(self.target_values) == 1
+        codegen(self.target_values[0])
+        codegen.extend_output(create_call_function(1, False))
```
Sorry, this still seems like it's creating a new stream?

voznesenskym replied:
See PR desc, this is a stream context, not a stream, and it was already doing this, but incorrectly.

ezyang commented on torch/_dynamo/variables/ctx_manager.py:
```diff
@@ -521,24 +521,28 @@ def exit(self, tx, *args):
         self.device = device
+        torch._dynamo.device_interface.register_interface_for_device(
+            self.device, torch._dynamo.device_interface.CudaInterface
+        )
```
What's this?

voznesenskym commented:
Back to drawing board, a few different things we want to try.
